apps/aefate/src/aefa_engine_state.erl

```diff
@@ -863,3 +923,88 @@ vm_version(#es{vm_version = X}) ->
     aetx_env:consensus_version(TxEnv).
+
+%%% Debugger functions
+
+-spec breakpoints(state()) -> list().
+breakpoints(#es{debug_info = #debug_info{breakpoints = Breakpoints}}) ->
+    Breakpoints.
+
+%%%------------------
+
+-spec skip_instructions(state()) -> integer().
+skip_instructions(#es{debug_info = #debug_info{skip_instructions = Skip}}) ->
+    Skip.
+
+-spec set_skip_instructions(integer(), state()) -> state().
+set_skip_instructions(Skip, ES = #es{debug_info = DbgInfo}) ->
+    ES#es{debug_info = DbgInfo#debug_info{skip_instructions = Skip}}.
+
+%%%------------------
+
+-spec debugger_status(state()) -> debugger_status() | disabled.
+debugger_status(#es{debug_info = disabled}) ->
+    disabled;
+debugger_status(#es{debug_info = #debug_info{status = Status}}) ->
+    Status.
+
+-spec set_debugger_status(debugger_status(), state()) -> state().
+set_debugger_status(Status, ES = #es{debug_info = DbgInfo}) ->
+    ES#es{debug_info = DbgInfo#debug_info{status = Status}}.
+
+%%%------------------
+
+-spec debugger_location(state()) -> debugger_location().
+debugger_location(#es{debug_info = #debug_info{location = Location}}) ->
+    Location.
+
+-spec set_debugger_location(debugger_location(), state()) -> state().
+set_debugger_location(Location, ES = #es{debug_info = DbgInfo}) ->
+    ES#es{debug_info = DbgInfo#debug_info{location = Location}}.
+
+%%%------------------
+
+-spec dbg_call_stack(state()) -> [{string(), pos_integer()}].
+dbg_call_stack(#es{debug_info = #debug_info{call_stack = Stack}}) ->
+    Stack.
+
+%%%------------------
+
+-spec add_variable_register(string(), tuple(), state()) -> state().
+add_variable_register(Var, Reg, ES = #es{debug_info = DbgInfo}) ->
+    VarsRegs = DbgInfo#debug_info.vars_registers,
+    Old = maps:get(Var, VarsRegs, []),
+    New = [Reg | Old],
+    NewDbgInfo = DbgInfo#debug_info{vars_registers = VarsRegs#{Var => New}},
+    ES#es{debug_info = NewDbgInfo}.
+
+-spec del_variable_register(string(), tuple(), state()) -> state().
+del_variable_register(Var, Reg, ES = #es{debug_info = DbgInfo}) ->
+    VarsRegs = DbgInfo#debug_info.vars_registers,
+    New = lists:delete(Reg, maps:get(Var, VarsRegs, [])),
+    NewDbgInfo = DbgInfo#debug_info{vars_registers = VarsRegs#{Var => New}},
+    ES#es{debug_info = NewDbgInfo}.
+
+-spec get_variable_register(string(), state()) -> tuple().
+get_variable_register(Var, #es{debug_info = #debug_info{vars_registers = VarsRegs}}) ->
+    case maps:get(Var, VarsRegs, [undefined]) of
+        []        -> undefined;
+        [Reg | _] -> Reg
+    end.
+
+%%%------------------
+
+debugger_resume(ES = #es{debug_info = Info = #debug_info{status = step}}) ->
+    ES#es{debug_info = Info#debug_info{status = break}};
+debugger_resume(ES = #es{debug_info = Info = #debug_info{status = {next, From}}}) ->
+    case ES#es.current_function == From of
+        true  -> ES#es{debug_info = Info#debug_info{status = break}};
+        false -> ES
+    end;
```

Comment:
This should be resolved using callstack --- otherwise you will always enter recursive calls (indirect ones too!)

Reply:
This should be resolved after this 74fd3e8dc98909c3590de895ec5153c35a03dcae